### DISCUSSION TITLE
@xstate/fsm and xstate supported features comparison

### DISCUSSION THREAD

**Discussion by yoav-lavi** [statelyai/xstate#1884] [Q&A] [answered]
Hey all!

First of all - XState is great, really looking forward to working with it.

Is there a detailed documentation on what features are supported by `xstate` but not by `@xstate/fsm`?

I'm working in a constrained environment where the full `xstate` wouldn't be practical (due to size constraints) but `@xstate/fsm` + `@xstate/react` (for `preact/compat`) are a good fit. I've been running into small things that don't seem to work on `@xstate/fsm` (which is completely understandable), and I'm trying to be able to both understand the differences completely and be able to avoid things in the documentation that are incompatible.

I've seen the README at https://github.com/davidkpiano/xstate/tree/master/packages/xstate-fsm and am using it, but I'm hoping to find a comparison flagging unsupported features if one exists (for example the README does mention "cond? (Guard)" which could be interpreted as supporting string guards and the object, but that hasn't worked on my end)

So far I've run into:

- The CodeSandbox visualizer doesn't seem to work for `@xstate/fsm` (I can use the normal `xstate` for that but that doesn't include throwing an error in cases where I've used something incompatible)
- `always` doesn't seem to be supported
- The guards property and string guards don't seem to work (using the function directly does)

Many thanks!

**Comment by yoav-lavi** [comment]
Updating after spending some time with `@xstate/fsm` and the README:

> The CodeSandbox visualizer doesn't seem to work for `@xstate/fsm` (I can use the normal `xstate` for that but that doesn't include throwing an error in cases where I've used something incompatible) 

This can be fixed by also creating an additional machine from the same configuration (only for the visualizer) from `xstate` and using the `@xstate/fsm` types:

```ts
import { Machine, interpret } from 'xstate';
import { inspect } from '@xstate/inspect';

const debugFsm = (config) => {
  inspect({ url: 'https://statecharts.io/inspect', iframe: false });
  const machine = Machine(config);
  interpret(machine, { devTools: true }).start();
};

export default debugFsm;
```

```ts
/* const machine = ... using the @xstate/fsm types here will also help detect use of incompatible features */
debugFsm(machine.config); // this will enable the visualizer
const service = interpret(machine); // this will also expect a correctly typed machine (for `@xstate/fsm`)
/* ... any use of `service` ... */
```

> `always` doesn't seem to be supported 

This doesn't seem to be mentioned in the `@xstate/fsm` documentation so it's just a matter of noticing that it's not supported / starting from the `@xstate/fsm` README directly rather than reading the `xstate` documentation first

> The `guards` property and string guards don't seem to work (using the function directly does)

This may just be a terminology difference between `xstate` and `@xstate/fsm`

It may still be beneficial to have a table of differences between the two but it looks like only taking the `@xstate/fsm` README into account is enough

**Reply by tomByrer** [reply]
I suggested before that it would be nice in the the main docs to flag if a feature was in FSM, just like they flag what version a new feature was added.

**Comment by davidkpiano** [comment, marked as answer]
Added a PR for this: https://github.com/davidkpiano/xstate/pull/1889/files Please comment in the PR and let me know if I missed anything:

|                             | **@xstate/fsm** | [XState](https://github.com/davidkpiano/xstate) |
| --------------------------- | :-------------: | :---------------------------------------------: |
| Finite states               |       ✅        |                       ✅                        |
| Initial state               |       ✅        |                       ✅                        |
| Transitions (object)        |       ✅        |                       ✅                        |
| Transitions (string target) |       ✅        |                       ✅                        |
| Delayed transitions         |       ❌        |                       ✅                        |
| Eventless transitions       |       ❌        |                       ✅                        |
| Nested states               |       ❌        |                       ✅                        |
| Parallel states             |       ❌        |                       ✅                        |
| History states              |       ❌        |                       ✅                        |
| Final states                |       ❌        |                       ✅                        |
| Context                     |       ✅        |                       ✅                        |
| Entry actions               |       ✅        |                       ✅                        |
| Exit actions                |       ✅        |                       ✅                        |
| Transition actions          |       ✅        |                       ✅                        |
| Parameterized actions       |       ❌        |                       ✅                        |
| Transition guards           |       ✅        |                       ✅                        |
| Parameterized guards        |       ❌        |                       ✅                        |
| Spawned actors              |       ❌        |                       ✅                        |
| Invoked actors              |       ❌        |                       ✅                        |

**Reply by yoav-lavi** [reply]
This is really helpful (and will hopefully help other people as well), thank you! 
I'll comment on the PR with a few questions / suggestions
